pkg/operator/server.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"k8s.io/component-base/cli/flag"
+)
+
+// TLSServerConfig contains the necessary fields to configure
+// web server TLS
+type TLSServerConfig struct {
+	CertFile       string
+	KeyFile        string
+	ClientCAFile   string
+	MinVersion     string
+	CipherSuites   []string
+	ReloadInterval time.Duration
+}
+
+// NewTLSConfig provides new server TLS configuration.
+func NewTLSConfig(logger log.Logger, certFile, keyFile, clientCAFile, minVersion string, cipherSuites []string) (*tls.Config, error) {
+	if certFile == "" && keyFile == "" {
+		if clientCAFile != "" {
+			return nil, errors.New("when a client CA is used a server key and certificate must also be provided")
+		}
+
+		level.Info(logger).Log("msg", "TLS disabled key and cert must be set to enable")
+
+		return nil, nil
+	}
+
+	level.Info(logger).Log("msg", "enabling server side TLS")
+
+	tlsCert, err := tls.LoadX509KeyPair(certFile, keyFile)
```

Comment:
the certificates are already loaded with the GetCertificate method, why do we need to load them explicitly here?

Reply:
You're  right, they will be loaded by the reloader later, although it might be useful to load them here during startup just to make sure they are valid files?

Reply:
We have no guarantee that that will continue to be the case, I'd rather have the behavior be consistent, and always log. Plus I think if a certificate is set the go TLS listener will not even attempt to load the certificate via the `GetCertificate` function.

Reply:
Ok, sounds good.  I think your suggestion also makes sense in case the secrets are not immediately mounted to the container.  I was hoping to activate TLS automatically without needing another flag by detecting if the relevant files are available, but it seems like it would be better to just have an explicit TLS activation flag, so I'll add that.

I tested that the listener will still reload even if the cert fields are set, but I agree it's better to just always reload the same way for consistency.

Reply:
Nice :+1: